protocols/request-response/src/cbor.rs

```diff
@@ -0,0 +1,229 @@
+// Copyright 2023 Protocol Labs
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::{Config, ProtocolSupport};
+use libp2p_swarm::NetworkBehaviour;
+use serde::{de::DeserializeOwned, Serialize};
+use std::ops::Deref;
+
+pub type OutEvent<Req, Resp> = crate::Event<Req, Resp>;
+
+#[derive(NetworkBehaviour)]
+#[behaviour(
+    to_swarm = "OutEvent<Req, Resp>",
+    prelude = "libp2p_swarm::derive_prelude"
+)]
+pub struct Behaviour<Req, Resp>
+where
+    Req: Send + Clone + Serialize + DeserializeOwned + 'static,
+    Resp: Send + Clone + Serialize + DeserializeOwned + 'static,
+{
+    inner: crate::Behaviour<codec::Codec<Req, Resp>>,
+}
+
+impl<Req, Resp> Behaviour<Req, Resp>
+where
+    Req: Send + Clone + Serialize + DeserializeOwned,
+    Resp: Send + Clone + Serialize + DeserializeOwned,
+{
+    pub fn new<I>(protocols: I, cfg: Config) -> Self
+    where
+        I: IntoIterator<
+            Item = (
+                <codec::Codec<Req, Resp> as crate::Codec>::Protocol,
+                ProtocolSupport,
+            ),
+        >,
+    {
+        Behaviour {
+            inner: crate::Behaviour::new(codec::Codec::default(), protocols, cfg),
+        }
+    }
+}
+
+impl<Req, Resp> Deref for Behaviour<Req, Resp>
+where
+    Req: Send + Clone + Serialize + DeserializeOwned,
+    Resp: Send + Clone + Serialize + DeserializeOwned,
+{
+    type Target = crate::Behaviour<codec::Codec<Req, Resp>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
```

Comment:
While this may be appealing, it is technically not correct. `Deref` should only be used for smart pointers (things like `Arc`, `Box`, `Rc`, etc). See https://doc.rust-lang.org/std/ops/trait.Deref.html.

The only reason we need the wrapping type is because we want a constructor that implicitly constructs the codec for us. I just had an idea on how we can solve this without `Deref`!

Reply:
Let me know what you think of [74669a3](https://github.com/libp2p/rust-libp2p/pull/3952/commits/74669a3b8a36a397d81765880cdc488fcefba96c).

Reply:
> Let me know what you think of [74669a3](https://github.com/libp2p/rust-libp2p/pull/3952/commits/74669a3b8a36a397d81765880cdc488fcefba96c).

We wanted to hide `codec` impls, but now we should know what `codec` to pass to `Behaviour::with_codec`. 
Isn't it?

Reply:
The next problem here is: we don't provide any information what type Req/Resp should be.
If in the example we remove 
```
#[derive(serde::Serialize, serde::Deserialize)]
```
Compiler would say nothing.

Reply:
> > Let me know what you think of [74669a3](https://github.com/libp2p/rust-libp2p/pull/3952/commits/74669a3b8a36a397d81765880cdc488fcefba96c).
> 
> We wanted to hide `codec` impls, but now we should know what `codec` to pass to `Behaviour::with_codec`. Isn't it?

But our `Codec`s also implement `Default`! Meaning users can use the `Behaviour::new` function and because the type-alias for `cbor::Behaviour` defines the codec, the `new` function knows which codec to instantiate.

Check the examples that I added :)

Reply:
Looks

> > > Let me know what you think of [74669a3](https://github.com/libp2p/rust-libp2p/pull/3952/commits/74669a3b8a36a397d81765880cdc488fcefba96c).
> > 
> > 
> > We wanted to hide `codec` impls, but now we should know what `codec` to pass to `Behaviour::with_codec`. Isn't it?
> 
> But our `Codec`s also implement `Default`! Meaning users can use the `Behaviour::new` function and because the type-alias for `cbor::Behaviour` defines the codec, the `new` function knows which codec to instantiate.
> 
> Check the examples that I added :)

Sorry. Yes, this works well.
Thanks!

I fixed your remark about `fn read_to_end`, please take a look.